conans/client/cmd/uploader.py

```diff
@@ -185,27 +211,49 @@ def _upload_ref(self, conanfile, ref, prefs, retry, retry_wait, integrity_check,
                                    reference=ref, remote=recipe_remote)
-
-        self._output.info("Uploading %s to remote '%s'" % (str(ref), recipe_remote.name))
+        msg = "\rUploading %s to remote '%s'" % (str(ref), recipe_remote.name)
+        self._output.info(left_justify_message(msg))
         self._upload_recipe(ref, conanfile, retry, retry_wait, policy, recipe_remote, remotes)
         upload_recorder.add_recipe(ref, recipe_remote.name, recipe_remote.url)
 
         # Now the binaries
         if prefs:
             total = len(prefs)
-            for index, pref in enumerate(prefs):
-                p_remote = recipe_remote
-                msg = ("Uploading package %d/%d: %s to '%s'" % (index+1, total, str(pref.id),
-                                                                p_remote.name))
-                self._output.info(msg)
-                self._upload_package(pref, retry, retry_wait,
-                                     integrity_check, policy, p_remote)
-                upload_recorder.add_package(pref, p_remote.name, p_remote.url)
-
-        # FIXME: I think it makes no sense to specify a remote to "post_upload"
-        # FIXME: because the recipe can have one and the package a different one
-        self._hook_manager.execute("post_upload", conanfile_path=conanfile_path, reference=ref,
-                                   remote=recipe_remote)
+            p_remote = recipe_remote
+
+            def upload_package_index(index_pref):
+                try:
+                    index, pref = index_pref
+                    up_msg = "\rUploading package %d/%d: %s to '%s'" % (index + 1, total,
+                                                                        str(pref.id),
+                                                                        p_remote.name)
+                    self._output.info(left_justify_message(up_msg))
+                    self._upload_package(pref, retry, retry_wait,
+                                         integrity_check, policy, p_remote)
+                    upload_recorder.add_package(pref, p_remote.name, p_remote.url)
+                    return conanfile_path, ref, recipe_remote, None
+                except ConanException as exc:
+                    return None, None, None, exc
+
+            def upload_package_callback(ret):
+                for cf_path, r_ref, r_rem, exc in ret:
+                    if exc is None:
+                        # FIXME: I think it makes no sense to specify a remote to "post_upload"
+                        # FIXME: because the recipe can have one and the package a different one
+                        self._hook_manager.execute("post_upload", conanfile_path=cf_path,
+                                                   reference=r_ref, remote=r_rem)
+                    else:
+                        self._exceptions_list.append(exc)
+
+            self._upload_thread_pool.map_async(upload_package_index,
```

Comment:
Add a comment: `This doesn't wait for the packages to end, so the function returns and the "pool entry" for the recipe is released`